### PR TITLE
feat(analytics): add pinpoint event submission

### PR DESCRIPTION
**Pull request by phantumcode**
*Issue #, if available:*

*Description of changes:*
* Initial implementation of saving local pinpoint events and submitting events to Pinpoint via pinpoint client
* Unit tests and open TODOs are in-progress/pending work yet to be completed.

*Check points: (check or cross out if not relevant)*

- [ ] ~Added new tests to cover change, if needed~
- [x] Build succeeds with all target using Swift Package Manager
- [ ] ~All unit tests pass~
- [ ] ~All integration tests pass~
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [] ~Documentation update for the change if required~
- [x] PR title conforms to conventional commit style
- [ ] ~If breaking change, documentation/changelog update with migration instructions~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/EventRecorder.swift [hunk 1]
```diff
@@ -5,21 +5,116 @@
...
+/// An AnalyticsEventRecording implementation that stores and submits pinpoint events
 class EventRecorder: AnalyticsEventRecording {
-    func save(_ event: PinpointEvent) async throws {
-        // TODO: Implement
-        fatalError("Not yet implemented")
+    let appId: String
+    let storage: AnalyticsEventStorage
+    let pinpointClient: PinpointClientProtocol
+    let AWSPinpointServiceDefinedMaxEventsPerBatch = 100
+    var submittedEvents: [PinpointEvent] = []
```
Comment (ruisebas) [inline AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/EventRecorder.swift:28]: `private`?

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/AnalyticsClient.swift [hunk 2]
```diff
@@ -7,17 +7,30 @@
...
 
 class AnalyticsClient: InternalPinpointClient {
-    private let eventRecorder: AnalyticsEventRecording
+    private var eventRecorder: AnalyticsEventRecording?
     private lazy var globalAttributes: [String: String] = [:]
     private lazy var globalMetrics: [String: Double] = [:]
     private lazy var eventTypeAttributes: [String: [String: String]] = [:]
     private lazy var eventTypeMetrics: [String: [String: Double]] = [:]
+    //needs to be this path to be backward compatible with legacy aws-ios-sdk
+    private let AWSPinpointClientRecorderDatabasePathPrefix = "com/amazonaws/AWSPinpointRecorder/"
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/AnalyticsClient.swift]: Done.

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/EventRecorder.swift [hunk 3]
```diff
@@ -5,21 +5,116 @@
...
 
+/// An AnalyticsEventRecording implementation that stores and submits pinpoint events
 class EventRecorder: AnalyticsEventRecording {
-    func save(_ event: PinpointEvent) async throws {
-        // TODO: Implement
-        fatalError("Not yet implemented")
+    let appId: String
+    let storage: AnalyticsEventStorage
+    let pinpointClient: PinpointClientProtocol
+    let AWSPinpointServiceDefinedMaxEventsPerBatch = 100
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/EventRecorder.swift]: Done.

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/EventRecorder.swift (hunk 1 above)
Comment (phantumcode) [reply AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/EventRecorder.swift]: updated var to be private

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/PinpointEvent+Bindings.swift [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import SQLite
+
+extension PinpointEvent {
+    static var archiver: AmplifyArchiverBehaviour {
+        return AmplifyArchiver()
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/PinpointEvent+Bindings.swift]: Good catch, fixed it.

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/PinpointEvent+Bindings.swift [hunk 5]
```diff
@@ -9,9 +9,10 @@ import Foundation
 import SQLite
 
 extension PinpointEvent {
-    static var archiver: AmplifyArchiverBehaviour {
-        return AmplifyArchiver()
-    }
+    typealias PintPointEventAttributes = [String: String]
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/PinpointEvent+Bindings.swift]: Great suggestion, I've moved it to AnalyticsClient.

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/AnalyticsClient.swift [hunk 6]
```diff
@@ -201,6 +201,8 @@ extension AnalyticsClient {
                 static let transactionId = "_transaction_id"
             }
         }
+        
+        static let EventRecorderStoragePathPrefix = "com/amazonaws/AWSPinpointRecorder/"
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Analytics/AnalyticsClient.swift]: I've updated the naming to be consistent.

**Review by ruisebas — APPROVED** [review]
:shipit:

**Pull request merged**